batch/batch/server/server.py

```diff
@@ -417,7 +417,39 @@ def to_json(self):
 log.info(f'app.root_path = {app.root_path}')
 
 
+authorized_users = [
+    'konradk@broadinstitute.org',
+    'labbott@broadinstitute.org',
+    'dking@broadinstitute.org',
+    'akotlar@broadinstitute.org',
+    'jigold@broadinstitute.org',
+    'ci'
```

Comment:
@akotlar How will our service accounts look?

Reply:
Plan was to name all service accounts by the user id (`sub` claim). This needs to be of course trasnfrormed because of the restricted dchartersets both google cloud and Kubernetes employ. Kube names (253 chars, `[a-zA-Z0-9-]*`: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/)

Google service accounts: Name is auto-generated, using a prefix that you supply, which I intended to be the same `sub` claim (user id). For a name `google-oauth2-kotlar-test`

```
{'name': 'projects/hail-vdc/serviceAccounts/google-oauth2-kotlar-test@hail-vdc.iam.gserviceaccount.com', 'projectId': 'hail-vdc', 'uniqueId': '101901669302330053293', 'email': 'google-oauth2-kotlar-test@hail-vdc.iam.gserviceaccount.com', 'etag': 'MDEwMjE5MjA=', 'oauth2ClientId': '101901669302330053293'}
```

Google Cloud service account characterset is slightly more restrictive than Kubernetes: (` [a-z][a-z\d\-]*[a-z\d]`) (no '.' allowed):

```
googleapiclient.errors.HttpError: <HttpError 400 when requesting https://iam.googleapis.com/v1/projects/hail-vdc/serviceAccounts?alt=json returned "google-oauth2|kotlar-test does not match [a-z][a-z\d\-]*[a-z\d].">
```

Reply:
Transformation can continue to be a hash of the user id, as done by notebook2 now (sha224, whose hex digest < 63 chars, which is the largest a label allows), or, if we trust that the user id is truly <provider>|<[a-Z0-9-]+, just replace the pipe with a dash. However, when I inquired about the guarantees, the answer I got suggested there is no guarantee: https://community.auth0.com/t/sub-claim-format-kubernetes-restricted-set/21996

...and the OpenID Connect spec that Auth0 follows specifies ASCII, which obviously means relying on Auth0 to use alphanumerics is fragile.

Reply:
A note on id uniqueness, stability: https://openid.net/specs/openid-connect-core-1_0.html#ClaimStability

Reply:
can we keep a mapping in a db somewhere from auth0 name to an int64 and use that everywhere as the user id? makes debugging annoying, but there seems to be a lot of issues that disappear if we use an id.

Reply:
@danking agreed.

Reply:
The solution I'm using at the moment is for service names.

user-<shortuuid> (22 character/byte uuid)

This id will get stored in the user's id token, which provides the mapping. We can of course maintain our own user db, but if we have the ability to store custom claims, is that not sufficient?

Reply:
I think he meant the same db that stores the other claims like bucket and service accounts.

 > The solution I'm using at the moment for service names

You mean k8s service names?  For maximum confusion, there are three kinds of service accounts: k8s service accounts, GCP service accounts, and Hail service accounts like ci.  Hail service accounts use Hail services like Hail users do, but are robots instead of users.  Hail service accounts need all the usual Hail users claims, like buckets and k8s and GCP service accounts, etc.

Reply:
For the initial approach, I am doing effectively:

```
create(username):
    sa_name = f'user-{shortuuid.uuid()}'

    out['sa_name'] = sa_name
    
    # Store username in 'displayName' property
    gsa_response = make_google_service_account(sa_name, username)
    out['gsa_unique_id'] = gsa_response['uniqueId']
    
    # Stores username in 'username' metadata
    ksa_response = make_kube_service_account(sa_name, username)
    
    out['ksa_name'] = sa_name

    out['bucket'] = create_bucket(sa_name)
```

The Google IAM response: 
```
{'name': 'projects/hail-vdc/serviceAccounts/user-14gwognesw8rw56xb3je73xgh@hail-vdc.iam.gserviceaccount.com', 'projectId': 'hail-vdc', 'uniqueId': '117029443340792843240', 'email': 'user-14gwognesw8rw56xb3je73xgh@hail-vdc.iam.gserviceaccount.com', 'displayName': 'akotlar', 'etag': 'MDEwMjE5MjA=', 'oauth2ClientId': '117029443340792843240'}
```

The gcloud output
```
$ gcloud iam service-accounts list
NAME                                    EMAIL
akotlar                                 user-02p21ibcb1axsokf1wp3vtcpd@hail-vdc.iam.gserviceaccount.com
```

The kube response
```
{'api_version': 'v1',
 'automount_service_account_token': None,
 'image_pull_secrets': None,
 'kind': 'ServiceAccount',
 'metadata': {'annotations': {'kubernetes.io/service-account.name': 'user-1a4zr260kqx8256mcg6mn3ue6',
                              'username': 'akotlar'},
              'cluster_name': None,
              'creation_timestamp': datetime.datetime(2019, 3, 15, 16, 56, 54, tzinfo=tzutc()),
              'deletion_grace_period_seconds': None,
              'deletion_timestamp': None,
              'finalizers': None,
              'generate_name': None,
              'generation': None,
              'initializers': None,
              'labels': None,
              'name': 'user-1a4zr260kqx8256mcg6mn3ue6',
              'namespace': 'default',
              'owner_references': None,
              'resource_version': '35974418',
              'self_link': '/api/v1/namespaces/default/serviceaccounts/user-1a4zr260kqx8256mcg6mn3ue6',
              'uid': '563c1dbe-4743-11e9-bcd4-42010a8000c9'},
 'secrets': None}
```

Reply:
Happy to go a different direction :)

Reply:
user-uuid sounds great.

I feel a bit weird about using a SHA (which I think is where the uuid comes from?) because it seems like we should be able to deterministically hand out int64's to users. I'm mostly ok with the uuid approach if it's substantially simpler.